### ISSUE TITLE
feat(trellis2-mv): native multi-image TRELLIS.2 pipeline (stochastic + multidiffusion)

### ISSUE THREAD

**Issue by raven38** [raven38/pixal3d.cpp#58]
## Goal

Implement a native C++/GGML **TRELLIS.2 multi-image / multiview image-to-3D pipeline** alongside the existing TRELLIS.2 single-image path and the separate Pixal3D MV path.

This is **multi-image conditioning without camera poses**. It must not reuse Pixal3D `--views` / `transforms.json` semantics.

### Reference state (2026-09-20)

- Microsoft TRELLIS.2 `main` exposes a single-image `Trellis2ImageTo3DPipeline.run(image, ...)`.
- microsoft/TRELLIS.2 PR #104 is an open multi-image implementation proposal using sampler injection with two modes:
  - `stochastic`: one conditioning image per diffusion/flow step, cycling through views.
  - `multidiffusion`: run the positive model prediction once per image at every step, average predictions, then apply the single negative prediction / CFG.
- Original TRELLIS also ships an experimental multi-image path with stochastic / multidiffusion.
- PR #104 discussion reports width/thickness drift under plain prediction averaging; confidence-weighted fusion is therefore a **post-MVP experiment**, not silently mixed into the reference-faithful baseline.

References:
- https://github.com/microsoft/TRELLIS.2/blob/main/trellis2/pipelines/trellis2_image_to_3d.py
- https://github.com/microsoft/TRELLIS.2/pull/104
- https://github.com/microsoft/TRELLIS/blob/main/app.py

## Target pipeline

```text
2..N images
  -> per-image single-image preprocessing (same contract as current TRELLIS.2)
  -> DINOv3 cond bank @512 (+ @1024 for cascade)
  -> MV sparse-structure flow fusion
  -> SS decode -> coords
  -> MV shape SLat fusion
       512 path
       OR 512 -> upsample -> 1024/1536 cascade
  -> MV texture SLat fusion
  -> existing shape/texture decode
  -> existing postprocess / GLB
```

No new model weights are required for the MVP; this is inference-time fusion over the existing TRELLIS.2 weights.

### stochastic

For sampler step `k`, use positive condition `cond[k % V]`. The negative condition remains the single zero condition. The view counter resets at the start of each sampler/stage, matching the reference proposal.

Expected cost: approximately single-image compute per step.

### multidiffusion

At each sampler step:

1. evaluate the positive prediction independently for every view;
2. average the **raw positive model predictions**;
3. evaluate the negative prediction once;
4. apply the existing CFG / guidance interval / guidance rescale semantics to the averaged positive prediction.

Do not average DINO tokens, final latents, coordinates, or meshes as a substitute.

Expected cost: approximately `V ×` flow-model compute for each guided stage.

## Compatibility / non-goals

- Existing TRELLIS.2 single-image `--image` behavior must remain unchanged.
- Existing Pixal3D SV/MV paths must remain unchanged.
- No camera transform estimation in this issue.
- No pose-aware projection / epipolar fusion.
- No specialized MV checkpoint training.
- No confidence-weighted fusion in the MVP.
- Do not change model manifests or require a second TRELLIS.2 model directory.

## User-facing contract

Proposed native CLI surface:

```text
--trellis2-mv DIR
--trellis2-mv-mode stochastic|multidiffusion
```

`DIR` is naturally sorted, contains 2..N image files, and is mutually exclusive with `--image`, `--sv-image`, and Pixal3D `--views`.

Server endpoint should be distinct from Pixal3D `/generate-mv`, e.g. `POST /generate-trellis2-mv`.

Studio should make TRELLIS.2 single-image vs TRELLIS.2 multiview explicit rather than silently changing behavior based on image count.

## Task sizing policy

Every implementation sub-issue below must be executable independently with an **agent context/work budget < 300k tokens**. If a task grows beyond that, split it before implementation.

Complexity scale:
- 1/5 trivial
- 2/5 small/local
- 3/5 moderate
- 4/5 cross-module / numerically sensitive
- 5/5 high-risk sampler/pipeline integration

## Dependency DAG

```text
reference oracle/fixtures
  ├─> multi-image preprocessing + DINO cond bank
  └─> generic MV sampler fusion
          │
cond bank ┴─> SS MV
                -> shape SLat MV/cascade
                    -> texture SLat MV
                        -> native CLI E2E
                            -> server API
                                -> Studio UI

native CLI E2E -------------------------> validation / parity / perf
server + Studio ------------------------^
validation -> confidence-weighted fusion follow-up (non-blocking)
```

## Parent acceptance

- [ ] All blocking sub-issues complete.
- [ ] 1-image regression demonstrates no behavior change to current TRELLIS.2.
- [ ] 2/4-view stochastic and multidiffusion run end-to-end at 1024.
- [ ] 512 and 1536 behavior is explicitly tested or intentionally gated.
- [ ] Stage-level results are compared against a pinned Python reference implementation.
- [ ] No NaN/Inf and no silent fallback to single-image behavior.
- [ ] CLI, server and Studio expose the mode explicitly.
- [ ] Linux CPU build + Web WASM regression remain green even when MV execution itself requires a capable backend.
- [ ] Performance/memory scaling with view count is documented.

## Sub-issues / execution order

| issue | task | complexity | agent budget | dependency |
|---|---|---:|---:|---|
| [ ] #59 | pinned Python reference + golden MV fixtures | 4/5 | 180k–240k | root |
| [ ] #60 | multi-image preprocessing + DINO cond bank | 3/5 | 120k–180k | #59 |
| [ ] #61 | generic stochastic / multidiffusion sampler | 5/5 | 220k–280k | #59 |
| [ ] #62 | SS flow + decode MV | 3/5 | 100k–160k | #60, #61 |
| [ ] #63 | shape SLat 512/1024/cascade MV | 5/5 | 220k–280k | #60, #61, #62 |
| [ ] #64 | texture SLat MV | 4/5 | 150k–220k | #60, #61, #63 |
| [ ] #65 | native CLI E2E + explicit flags | 4/5 | 160k–220k | #60, #62, #63, #64 |
| [ ] #66 | server `/generate-trellis2-mv` | 3/5 | 120k–180k | #65 |
| [ ] #67 | Studio explicit TRELLIS.2 MV mode | 4/5 | 160k–220k | #66 |
| [ ] #68 | final parity / quality / perf / memory gate | 5/5 | 220k–280k | #65, #66, #67 |
| [ ] #69 | confidence-weighted fusion research | 4/5 | 160k–240k | #68; **non-blocking** |

### Concrete DAG

```text
#59 reference
 ├──> #60 cond bank ───────────────┐
 └──> #61 sampler ───────────────┐ │
                                 v v
                                #62 SS
                                  |
                                  v
                                #63 shape
                                  |
                                  v
                                #64 texture
                                  |
                                  v
                                #65 CLI E2E
                                  |
                                  v
                                #66 server
                                  |
                                  v
                                #67 Studio
                                  |
                                  v
                                #68 final validation
                                  |
                                  v
                                #69 confidence weighting (non-blocking)
```

### Parallelism

After #59 completes:
- #60 and #61 can run in parallel.
- #62 starts when both finish.
- #68 stage-level parity work may start incrementally as #62/#63/#64 land, but final closure waits for #65/#66/#67.

All blocking tasks are sized below the 300k-token per-agent target.

**Comment by raven38** [comment]
Progress update (2026-09-21):

- #59: implementation/protocol is now in draft PR #70.
  - PR #104 reference SHA pinned.
  - weightless stochastic/multidiffusion oracle is in Linux CI and green.
  - full 512/1024-cascade fixture generator is implemented.
  - #59 remains open until real 1/2/4-view GPU fixtures + repeat-hash stability are recorded.
- #61: remains blocked by #59. Do not start numerical sampler integration until the real reference manifests are available.
- #60 can likewise use #70's cond dump contract once #59's real fixtures land.

Current critical path: #70 -> GPU reference fixture runs -> close #59 -> #60 and #61 in parallel.

**Comment by raven38** [comment]
## Clarification: "each sampler/stage" — the reset boundary is the **sampler**, not the cascade stage

The "Required fusion semantics → stochastic" section currently says:

> The view counter resets at the start of each sampler/stage, matching the reference proposal.

Checking microsoft/TRELLIS.2 PR #104 directly while pinning the reference for #59: the "/stage" half
of that phrase is misleading. In PR #104, `inject_sampler_multi_image()` is a context manager that
creates `counter = {'value': 0}` once on context entry, and the cascade branch wraps the entire
`sample_shape_slat_cascade(512 → target_res)` call in a **single** `with` block. The counter therefore
continues across the LR(512)→HR(1024/1536) boundary and only resets at the three sampler entries
(`sparse_structure_sampler`, `shape_slat_sampler`, `tex_slat_sampler`).

**Decision (repo owner, 2026-09-21): follow PR #104 literally.** The sentence should be read as:

> The view counter resets at the start of each **sampler** (three times: sparse-structure, shape SLat,
> texture SLat). Within the shape SLat cascade it **continues** across the LR→HR boundary.

#63's acceptance item "stochastic counter resets independently for LR and HR samplers" contradicted
this and has been corrected in #63.

Generated with Claude Code

**Comment by raven38** [comment]
## Baseline correctness exception to "single-image behavior must remain unchanged"

Adversarial verification of #61 found that the **inherited trellis.cpp `sample_flow`** computes the
flow t-schedule and the guidance-interval test in `float32`
(`src/flow_runner.cpp:395,398,429` at the epic base `24adff1`), whereas the pinned PyTorch reference
does both in `float64`:

- `trellis2/pipelines/samplers/flow_euler.py:115-117` — `t_seq` built with numpy float64 and
  `.tolist()`'d to Python floats
- `trellis2/pipelines/samplers/guidance_interval_mixin.py:9` — `guidance_interval[0] <= t <= guidance_interval[1]`
  compared as Python floats
- `flow_euler.py:45` — the `float32` cast (`torch.tensor([1000 * t], dtype=torch.float32)`) happens
  **after** the interval test

With the shape-SLat production parameters (`steps=12, rescale_t=3.0, guidance_interval=[0.6, 1.0]`),
step `i=8` lands **exactly** on `t = 0.6`; float64 classifies it inside the interval, float32 outside.
Per-stage counts (reference / native, inside/outside):

| stage | reference | native | effect |
|---|---|---|---|
| sparse-structure (`rescale_t=5.0`) | 10/2 | 10/2 | none |
| **shape SLat** | **9/3** | **8/4** | CFG on/off flips for that step |
| texture SLat | 6/6 | 5/7 | none (`gs=1.0, gr=0.0` → no-op) |

The magnitude on the real model is **not yet measured** (the only full-trajectory test with these
parameters, `trellis-test-slat-shape`, needs CUDA and the shape weights; it previously passed on a
CUDA machine within its `<3e-2` relative tolerance, so the effect is expected to be small but real).

**Decision (repo owner, 2026-09-21): fix the baseline.** The t-schedule and interval test move to a
shared helper used by both `sample_flow` and the new `sample_flow_multi` (so `V=1` stays bit-identical
by construction), reproducing numpy's `linspace` arithmetic and casting to `float32` at the same point
as the reference. This lands as a separate `fix(flow):` commit ahead of #61.

Consequently the non-goal *"Existing TRELLIS.2 single-image `--image` behavior must remain unchanged"*
is read as **unchanged except for this correctness fix**: single-image shape-SLat output may change at
step `i=8` (toward the reference). Re-running `trellis-test-slat-shape` on a CUDA pod before the
epic's final gate (#68) is required to quantify it.

Generated with Claude Code

**Comment by raven38** [comment]
## B3 (4-view stochastic texture black-out): closed as a documented limitation

Investigated on both sides and **not fixed**. Full record:
`docs/results/2026-09-21-trellis2-mv-b3/2026-09-22-b3-conclusion.md` (on `feat/trellis2-mv`).

### The facts

**The reference implementation blackens too.** The existing v3 fixture
`run_2img_real_stochastic_1024c` (2-view stochastic, seed 42, PyTorch pinned `75fbf018`) bakes a
texture with **every texel 0**, while its `tex_slat` is finite and in range (mean 0.29066333,
std 3.5458083) — so the black-out happens after decode. It is bit-identical to the seed-42 run of a
23-seed reference sweep (`ss_coords` / `tex_slat_{feats,coords}` sha256 all match). This also
refutes the premise the investigation started from, that the reference's stochastic runs were all
normal at their single seed.

**Fisher exact, two-sided** (black-rate, reference vs native):

| Comparison | reference | native | p |
|---|---|---|---|
| 4v stochastic | 0/10 | 2/6 | **0.1250** |
| 2v stochastic | 1/10 | 1/3 | **0.4231** |
| pooled (descriptive) | 1/20 | 3/9 | **0.0760** |
| sensitivity, `mean < 10` counted as not-normal — 4v | 2/10 | 2/6 | 0.6044 |
| same, 2v | 2/10 | 1/3 | 1.0000 |
| same, pooled | 4/20 | 3/9 | 0.6424 |

All six p-values were independently recomputed from the raw sweep JSONL by direct hypergeometric
summation and match the sweep report to four digits.

### What this does and does not say

**The reference does it too, and at n = 10 (reference) / 6 (native) no difference in rate was
detectable. There is no evidence of a native-specific bug.**

It is **not** "does not reproduce" and **not** "no problem":

- n = 10/6 cannot detect a difference — this is not proof the rates are equal.
- The 4-view point estimate is **higher for native** (0/10 vs 2/6, p = 0.1250).
- A native-specific contribution is **not excluded**: the tex DiT real-weight forward diverges from
  the reference by **1.29 % at step 0** (f16 GGUF; 5.12 % at step 11). Filed separately — it may be
  common to the single-image path, so it is not scoped to this epic.
- The decisive experiment — injecting the native noise into the reference so both follow the same
  trajectory — was **not run**: the tensor it needs (`tex_shape_guide_norm.npy`) was never dumped.

### Observations recorded without interpretation

Saturation rate separates the one fully-black run (0.9999911) but not the rest: three boundary runs
sit at 0.3975 / 0.4898 / 0.5007 against a normal range of 0.0–0.3675 for the other 19 — essentially
contiguous. Multidiffusion did not blacken in any run, but that is n = 1 native / 3 reference, which
is far too small to call it robust.

The reference sweep and calibration were **1024-cascade only**; the 512 fixtures were never
measured, so whether the black-out is specific to the 1024 cascade is untested. Reading those
existing GLBs locally would be the cheapest next piece of evidence.

### Workaround

Re-rolling the seed, or running `--trellis2-mv-mode multidiffusion`, which averages the views'
predictions instead of cycling one per step.

Generated with Claude Code
